core/src/Cuda/Kokkos_Cuda_Vectorization.hpp

```diff
@@ -70,16 +70,28 @@ struct in_place_shfl_op {
     return *static_cast<Derived const*>(this);
   }
 
-  // sizeof(Scalar) == sizeof(int) case
+  // sizeof(Scalar) <= sizeof(int) case
   template <class Scalar>
   // requires _assignable_from_bits<Scalar>
-  __device__ inline typename std::enable_if<sizeof(Scalar) == sizeof(int)>::type
+  __device__ inline typename std::enable_if<sizeof(Scalar) <= sizeof(int)>::type
   operator()(Scalar& out, Scalar const& in, int lane_or_delta, int width,
              unsigned mask = shfl_all_mask) const noexcept {
+    using shfl_type = int;
+    union conv_type {
+      Scalar orig;
+      shfl_type conv;
+    };
+    conv_type tmp_in;
```

Comment:
I'd prefer to add a `tmp_in.conv = 0` here in case there are any shenanigans with carry bits or something. Also to keep memory sanitizers happy, since we'd be doing things on uninitialized bits otherwise.